File: esp32/src/scheduler.cpp
#include "scheduler.h"
#include "ot_master.h"
#include "config_manager.h"
#include "wolf_cwl.h"
#include "logging.h"
#include "display.h"
#include <Preferences.h>
#include <ArduinoJson.h>
#include <time.h>

// =============================================================================
// Global state
// =============================================================================
ScheduleEntry schedules[MAX_SCHEDULES];
int scheduleCount = 0;
BypassSchedule bypassSchedule;

bool scheduleActive = false;
bool scheduleOverride = false;
bool bypassScheduleActive = false;
bool bypassOverride = false;

// Timed off state
bool timedOffActive = false;
time_t timedOffEndEpoch = 0;
uint8_t timedOffHours = 0;
static bool timedOffRestoredFromNvs = false;  // Deferred NTP check pending

static int lastActiveScheduleIndex = -1;
static bool lastBypassSeason = false;
static unsigned long lastEvalTime = 0;
static bool bypassSeasonInitialized = false;

#define EVAL_INTERVAL 60000  // Evaluate once per minute

static const char* PREFS_NAMESPACE = "wolfcwl";

// =============================================================================
// NVS persistence — Ventilation schedules
// =============================================================================
void loadSchedules() {
    Preferences prefs;
    prefs.begin(PREFS_NAMESPACE, true);
    scheduleCount = prefs.getInt("sched_count", 0);
    if (scheduleCount > MAX_SCHEDULES) scheduleCount = MAX_SCHEDULES;

    for (int i = 0; i < scheduleCount; i++) {
        String p = "sched_" + String(i) + "_";
        schedules[i].enabled = prefs.getBool((p + "en").c_str(), false);
        schedules[i].startHour = prefs.getInt((p + "sh").c_str(), 0);
        schedules[i].startMinute = prefs.getInt((p + "sm").c_str(), 0);
        schedules[i].endHour = prefs.getInt((p + "eh").c_str(), 0);
        schedules[i].endMinute = prefs.getInt((p + "em").c_str(), 0);
        schedules[i].ventLevel = prefs.getInt((p + "lv").c_str(), 2);
        schedules[i].activeDays = prefs.getInt((p + "dy").c_str(), 0x7F);  // All days
    }
    prefs.end();

    if (scheduleCount > 0) {
        log("Scheduler: Loaded " + String(scheduleCount) + " ventilation schedules");
    }
}

void saveSchedules() {
    Preferences prefs;
    prefs.begin(PREFS_NAMESPACE, false);
    prefs.putInt("sched_count", scheduleCount);

    for (int i = 0; i < MAX_SCHEDULES; i++) {
        String p = "sched_" + String(i) + "_";
        if (i < scheduleCount) {
            prefs.putBool((p + "en").c_str(), schedules[i].enabled);
            prefs.putInt((p + "sh").c_str(), schedules[i].startHour);
            prefs.putInt((p + "sm").c_str(), schedules[i].startMinute);
            prefs.putInt((p + "eh").c_str(), schedules[i].endHour);
            prefs.putInt((p + "em").c_str(), schedules[i].endMinute);
            prefs.putInt((p + "lv").c_str(), schedules[i].ventLevel);
            prefs.putInt((p + "dy").c_str(), schedules[i].activeDays);
        } else if (prefs.isKey((p + "en").c_str())) {
            prefs.remove((p + "en").c_str());
            prefs.remove((p + "sh").c_str());
            prefs.remove((p + "sm").c_str());
            prefs.remove((p + "eh").c_str());
            prefs.remove((p + "em").c_str());
            prefs.remove((p + "lv").c_str());
            prefs.remove((p + "dy").c_str());
        }
    }
    prefs.end();
    log("Scheduler: Saved " + String(scheduleCount) + " ventilation schedules");
}

// =============================================================================
// NVS persistence — Bypass schedule
// =============================================================================
void loadBypassSchedule() {
    Preferences prefs;
    prefs.begin(PREFS_NAMESPACE, true);
    bypassSchedule.enabled = prefs.getBool("bp_sched_en", false);
    bypassSchedule.startDay = prefs.getInt("bp_sched_sd", 15);
    bypassSchedule.startMonth = prefs.getInt("bp_sched_sm", 4);  // April
    bypassSchedule.endDay = prefs.getInt("bp_sched_ed", 30);
    bypassSchedule.endMonth = prefs.getInt("bp_sched_em", 9);    // September
    prefs.end();

    if (bypassSchedule.enabled) {
        log("Scheduler: Bypass schedule loaded (" +
            String(bypassSchedule.startDay) + "." + String(bypassSchedule.startMonth) +
            " - " + String(bypassSchedule.endDay) + "." + String(bypassSchedule.endMonth) + ")");
    }
}

void saveBypassSchedule() {
    Preferences prefs;
    prefs.begin(PREFS_NAMESPACE, false);
    prefs.putBool("bp_sched_en", bypassSchedule.enabled);
    prefs.putInt("bp_sched_sd", bypassSchedule.startDay);
    prefs.putInt("bp_sched_sm", bypassSchedule.startMonth);
    prefs.putInt("bp_sched_ed", bypassSchedule.endDay);
    prefs.putInt("bp_sched_em", bypassSchedule.endMonth);
    prefs.end();
    log("Scheduler: Bypass schedule saved");
}

// =============================================================================
// Setup
// =============================================================================
void setupScheduler() {
    memset(schedules, 0, sizeof(schedules));
    memset(&bypassSchedule, 0, sizeof(bypassSchedule));
    loadSchedules();
    loadBypassSchedule();

    // Restore persisted override state (written only on manual user actions)
    Preferences prefs;
    prefs.begin(PREFS_NAMESPACE, true);

    // Timed off
    uint32_t storedEnd = prefs.getUInt("toff_end", 0);
    if (storedEnd > 0) {
        timedOffEndEpoch = (time_t)storedEnd;
        timedOffActive = true;
        timedOffRestoredFromNvs = true;  // Will verify against NTP in schedulerLoop
        requestedVentLevel = VENT_LEVEL_OFF;
        scheduleOverride = true;
        log("Scheduler: Restored timed off from NVS (end=" + String(storedEnd) + ")");
    }

    // Manual overrides (only if not already in timed off)
    if (!timedOffActive) {
        scheduleOverride = prefs.getBool("sched_ovr", false);
        if (scheduleOverride) {
            log("Scheduler: Restored ventilation manual override from NVS");
        }
    }
    bypassOverride = prefs.getBool("bp_ovr", false);
    if (bypassOverride) {
        log("Scheduler: Restored bypass manual override from NVS");
    }

    prefs.end();
}

// =============================================================================
// Time matching helpers
// =============================================================================
static int timeToMinutes(uint8_t h, uint8_t m) {
    return h * 60 + m;
}

// Convert tm_wday (0=Sun) to our bitmask (bit0=Mon)
static uint8_t tmWdayToBitmask(int wday) {
    // wday: 0=Sun, 1=Mon, ..., 6=Sat
    // our bitmask: bit0=Mon, bit1=Tue, ..., bit5=Sat, bit6=Sun
    if (wday == 0) return (1 << 6);  // Sunday = bit 6
    return (1 << (wday - 1));
}

static bool isTimeInRange(int currentMinutes, int startMinutes, int endMinutes) {
    if (startMinutes <= endMinutes) {
        // Normal range (e.g., 08:00 - 22:00)
        return currentMinutes >= startMinutes && currentMinutes < endMinutes;
    } else {
        // Spans midnight (e.g., 22:00 - 06:00)
        return currentMinutes >= startMinutes || currentMinutes < endMinutes;
    }
}

// Check if a date (day, month) falls within [startDay.startMonth, endDay.endMonth]
static bool isDateInRange(int day, int month, int sd, int sm, int ed, int em) {
    int current = month * 100 + day;
    int start = sm * 100 + sd;
    int end = em * 100 + ed;

    if (start <= end) {
        // Normal range (e.g., Apr 15 - Sep 30)
        return current >= start && current <= end;
    } else {
        // Spans year boundary (e.g., Oct 1 - Mar 31)
        return current >= start || current <= end;
    }
}

// =============================================================================
// Ventilation schedule evaluation
// =============================================================================
static int evaluateVentilationSchedule(struct tm* timeinfo) {
    int currentMinutes = timeToMinutes(timeinfo->tm_hour, timeinfo->tm_min);
    uint8_t dayBit = tmWdayToBitmask(timeinfo->tm_wday);

    for (int i = 0; i < scheduleCount; i++) {
        if (!schedules[i].enabled) continue;
        if (!(schedules[i].activeDays & dayBit)) continue;

        int startMin = timeToMinutes(schedules[i].startHour, schedules[i].startMinute);
        int endMin = timeToMinutes(schedules[i].endHour, schedules[i].endMinute);

        if (isTimeInRange(currentMinutes, startMin, endMin)) {
            return i;
        }
    }
    return -1;  // No match
}

// =============================================================================
// Bypass schedule evaluation
// =============================================================================
static bool evaluateBypassSchedule(struct tm* timeinfo) {
    if (!bypassSchedule.enabled) return false;

    int day = timeinfo->tm_mday;
    int month = timeinfo->tm_mon + 1;  // tm_mon is 0-based

    return isDateInRange(day, month,
                         bypassSchedule.startDay, bypassSchedule.startMonth,
                         bypassSchedule.endDay, bypassSchedule.endMonth);
}

// =============================================================================
// Main scheduler loop
// =============================================================================
static void persistOverrideState() {
    Preferences prefs;
    prefs.begin(PREFS_NAMESPACE, false);
    prefs.putUInt("toff_end", timedOffActive ? (uint32_t)timedOffEndEpoch : 0);
    prefs.putBool("sched_ovr", scheduleOverride && !timedOffActive);
    prefs.putBool("bp_ovr", bypassOverride);
    prefs.end();
}

void schedulerLoop() {
    unsigned long now = millis();

    // --- Timed off: NTP-deferred validation after reboot ---
    time_t now_time = time(nullptr);
    bool ntpSynced = now_time > 1700000000;

    if (timedOffActive && timedOffRestoredFromNvs && ntpSynced) {
        timedOffRestoredFromNvs = false;
        if (timedOffEndEpoch <= now_time) {
            // Timer already expired during downtime
            log("Scheduler: Timed off expired during downtime, resuming Normal");
            timedOffActive = false;
            timedOffHours = 0;
            timedOffEndEpoch = 0;
            requestedVentLevel = VENT_LEVEL_NORMAL;
            scheduleOverride = false;
            persistOverrideState();
            displayWake();
        } else {
            timedOffHours = (uint8_t)((timedOffEndEpoch - now_time) / 3600 + 1);
            log("Scheduler: Timed off restored, " + String(timedOffHours) + "h remaining");
        }
    }

    // --- Timed off expiry check (epoch-based) ---
    if (timedOffActive && !timedOffRestoredFromNvs && ntpSynced && now_time >= timedOffEndEpoch) {
        log("Scheduler: Timed off expired, resuming Normal");
        timedOffActive = false;
        timedOffHours = 0;
        timedOffEndEpoch = 0;
        requestedVentLevel = VENT_LEVEL_NORMAL;
        scheduleOverride = false;
        persistOverrideState();
        displayWake();
    }

    if (now - lastEvalTime < EVAL_INTERVAL) return;
    lastEvalTime = now;

    if (!ntpSynced) return;

    struct tm timeinfo;
    localtime_r(&now_time, &timeinfo);

    // --- Ventilation schedule (skip if timed off is active) ---
    if (!timedOffActive) {
        int matchIndex = evaluateVentilationSchedule(&timeinfo);

        if (matchIndex != lastActiveScheduleIndex) {
            if (scheduleOverride) {
                scheduleOverride = false;
                log("Scheduler: Override cleared (schedule transition)");
            }
            lastActiveScheduleIndex = matchIndex;
        }

        if (!scheduleOverride) {
            if (matchIndex >= 0) {
                scheduleActive = true;
                uint8_t level = schedules[matchIndex].ventLevel;
                if (level != requestedVentLevel) {
                    requestedVentLevel = level;
                    displayWake();
                    log("Scheduler: Level set to " + String(level) +
                        " (" + String(getVentilationLevelName(level)) + ")");
                }
            } else {
                if (scheduleActive) {
                    scheduleActive = false;
                    requestedVentLevel = appConfig.ventilationLevel;
                    displayWake();
                    log("Scheduler: No schedule active, using default level " +
                        String(appConfig.ventilationLevel));
                }
            }
        }
    }

    // --- Bypass schedule ---
    bool isSummer = evaluateBypassSchedule(&timeinfo);

    if (bypassSeasonInitialized && isSummer != lastBypassSeason) {
        // Season transition — clear override
        if (bypassOverride) {
            bypassOverride = false;
            log("Scheduler: Bypass override cleared (season transition)");
        }
    }
    lastBypassSeason = isSummer;
    bypassSeasonInitialized = true;

    if (bypassSchedule.enabled && !bypassOverride) {
        bypassScheduleActive = true;
        if (isSummer != requestedBypassOpen) {
            requestedBypassOpen = isSummer;
            displayWake();
            log("Scheduler: Bypass " + String(isSummer ? "open (summer)" : "closed (winter)"));
        }
    } else if (!bypassSchedule.enabled) {
        bypassScheduleActive = false;
    }
}

// =============================================================================
// Manual override
// =============================================================================
void setVentilationManualOverride() {
    if (scheduleActive) {
        scheduleOverride = true;
        persistOverrideState();
        log("Scheduler: Ventilation manual override set (persisted)");
    }
}

void setBypassManualOverride() {
    if (bypassScheduleActive) {
        bypassOverride = true;
        persistOverrideState();
        log("Scheduler: Bypass manual override set (persisted)");
    }
}

// =============================================================================
// Timed off mode
// =============================================================================
void activateTimedOff(uint8_t hours) {
    if (hours == 0 || hours > 99) return;
    time_t now_time = time(nullptr);
    if (now_time < 1700000000) return;  // Need NTP for epoch

    timedOffActive = true;
    timedOffHours = hours;
    timedOffEndEpoch = now_time + (time_t)hours * 3600;
    timedOffRestoredFromNvs = false;
    requestedVentLevel = VENT_LEVEL_OFF;
    scheduleOverride = true;
    persistOverrideState();
    displayWake();
    log("Scheduler: Timed off activated for " + String(hours) + "h (persisted)");
}

void cancelTimedOff() {
    if (!timedOffActive) return;
    timedOffActive = false;
    timedOffHours = 0;
    timedOffEndEpoch = 0;
    timedOffRestoredFromNvs = false;
    requestedVentLevel = VENT_LEVEL_NORMAL;
    scheduleOverride = false;
    persistOverrideState();
    displayWake();
    log("Scheduler: Timed off cancelled, resuming Normal (persisted)");
}

unsigned long getTimedOffRemainingMinutes() {
    if (!timedOffActive) return 0;
    time_t now_time = time(nullptr);
    if (now_time < 1700000000) return 0;  // NTP not synced
    if (now_time >= timedOffEndEpoch) return 0;
    return (unsigned long)(timedOffEndEpoch - now_time) / 60;
}

// =============================================================================
// JSON serialization — Ventilation schedules
// =============================================================================
String getSchedulesJson() {
    JsonDocument doc;
    JsonArray arr = doc.to<JsonArray>();

    for (int i = 0; i < scheduleCount; i++) {
        JsonObject entry = arr.add<JsonObject>();
        entry["enabled"] = schedules[i].enabled;
        entry["startHour"] = schedules[i].startHour;
        entry["startMinute"] = schedules[i].startMinute;
        entry["endHour"] = schedules[i].endHour;
        entry["endMinute"] = schedules[i].endMinute;
        entry["ventLevel"] = schedules[i].ventLevel;
        entry["activeDays"] = schedules[i].activeDays;
    }

    String output;
    serializeJson(doc, output);
    return output;
}

bool updateSchedulesFromJson(const String& json) {
    JsonDocument doc;
    DeserializationError error = deserializeJson(doc, json);
    if (error) {
        log("Scheduler: JSON parse error: " + String(error.c_str()));
        return false;
    }

    if (!doc.is<JsonArray>()) return false;
    JsonArray arr = doc.as<JsonArray>();

    scheduleCount = 0;
    for (JsonObject entry : arr) {
        if (scheduleCount >= MAX_SCHEDULES) break;
        int i = scheduleCount;
        schedules[i].enabled = entry["enabled"] | false;
        schedules[i].startHour = entry["startHour"] | 0;
        schedules[i].startMinute = entry["startMinute"] | 0;
        schedules[i].endHour = entry["endHour"] | 0;
        schedules[i].endMinute = entry["endMinute"] | 0;
        schedules[i].ventLevel = entry["ventLevel"] | 2;
        schedules[i].activeDays = entry["activeDays"] | 0x7F;
        scheduleCount++;
    }

    saveSchedules();
    return true;
}

// =============================================================================
// JSON serialization — Bypass schedule
// =============================================================================
String getBypassScheduleJson() {
    JsonDocument doc;
    doc["enabled"] = bypassSchedule.enabled;
    doc["startDay"] = bypassSchedule.startDay;
    doc["startMonth"] = bypassSchedule.startMonth;
    doc["endDay"] = bypassSchedule.endDay;
    doc["endMonth"] = bypassSchedule.endMonth;

    String output;
    serializeJson(doc, output);
    return output;
}

bool updateBypassScheduleFromJson(const String& json) {
    JsonDocument doc;
    DeserializationError error = deserializeJson(doc, json);
    if (error) return false;

    bypassSchedule.enabled = doc["enabled"] | false;
    bypassSchedule.startDay = doc["startDay"] | 15;
    bypassSchedule.startMonth = doc["startMonth"] | 4;
    bypassSchedule.endDay = doc["endDay"] | 30;
    bypassSchedule.endMonth = doc["endMonth"] | 9;

    saveBypassSchedule();
    return true;
}

// =============================================================================
// Active schedule description
// =============================================================================
static const char* daysLabel(uint8_t mask) {
    if (mask == 0x7F) return "Every day";
    if (mask == 0x1F) return "Weekdays";
    if (mask == 0x60) return "Weekend";
    return nullptr;  // Use individual days
}

String getActiveScheduleDescription() {
    if (!scheduleActive || lastActiveScheduleIndex < 0) return "";

    ScheduleEntry& s = schedules[lastActiveScheduleIndex];
    const char* days = daysLabel(s.activeDays);
    char buf[64];
    if (days) {
        snprintf(buf, sizeof(buf), "%02d:%02d-%02d:%02d %s %s",
                 s.startHour, s.startMinute, s.endHour, s.endMinute,
                 getVentilationLevelName(s.ventLevel), days);
    } else {
        snprintf(buf, sizeof(buf), "%02d:%02d-%02d:%02d %s",
                 s.startHour, s.startMinute, s.endHour, s.endMinute,
                 getVentilationLevelName(s.ventLevel));
    }
    return String(buf);
}

File: esp32/src/scheduler.rs
//! Ventilation and bypass scheduling with manual override and timed-off support.

use std::sync::{Arc, Mutex};
use log::info;
use serde::{Deserialize, Serialize};

use crate::app_state::AppStateInner;

type AppState = Arc<Mutex<AppStateInner>>;

#[allow(dead_code)]
pub const MAX_SCHEDULES: usize = 16;
const EVAL_INTERVAL_MS: u32 = 60_000;

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ScheduleEntry {
    pub enabled: bool,
    #[serde(rename = "startHour")]
    pub start_hour: u8,
    #[serde(rename = "startMinute")]
    pub start_minute: u8,
    #[serde(rename = "endHour")]
    pub end_hour: u8,
    #[serde(rename = "endMinute")]
    pub end_minute: u8,
    #[serde(rename = "ventLevel")]
    pub vent_level: u8,
    #[serde(rename = "activeDays")]
    pub active_days: u8, // bitmask: bit0=Mon..bit6=Sun
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct BypassSchedule {
    pub enabled: bool,
    #[serde(rename = "startDay")]
    pub start_day: u8,
    #[serde(rename = "startMonth")]
    pub start_month: u8,
    #[serde(rename = "endDay")]
    pub end_day: u8,
    #[serde(rename = "endMonth")]
    pub end_month: u8,
}

pub struct Scheduler {
    state: AppState,
    pub schedules: Vec<ScheduleEntry>,
    pub bypass_schedule: BypassSchedule,
    last_eval_ms: u32,
    last_active_index: i32,
    pub timed_off_active: bool,
    pub timed_off_end_epoch: i64,
}

impl Scheduler {
    pub fn new(state: AppState) -> Self {
        Self {
            state,
            schedules: Vec::new(),
            bypass_schedule: BypassSchedule::default(),
            last_eval_ms: 0,
            last_active_index: -1,
            timed_off_active: false,
            timed_off_end_epoch: 0,
        }
    }

    pub fn update(&mut self, now_ms: u32) {
        // Check timed off expiration
        if self.timed_off_active {
            let now_epoch = unsafe { esp_idf_svc::sys::time(std::ptr::null_mut()) } as i64;
            if now_epoch > 1_700_000_000 && now_epoch >= self.timed_off_end_epoch {
                info!("Scheduler: Timed off expired");
                self.timed_off_active = false;
                self.timed_off_end_epoch = 0;
                let mut st = self.state.lock().unwrap();
                st.requested_vent_level = 2; // Normal
                st.schedule_override = false;
                st.timed_off_end_epoch = 0;
                st.persist_timed_off = true;
            }
        }

        // Update timed-off status in shared state
        {
            let mut st = self.state.lock().unwrap();
            st.timed_off_active = self.timed_off_active;
            st.timed_off_remaining_min = self.timed_off_remaining_minutes();
        }

        if now_ms.wrapping_sub(self.last_eval_ms) < EVAL_INTERVAL_MS {
            return;
        }
        self.last_eval_ms = now_ms;

        // Sync schedules from AppState (may have been updated via web UI)
        {
            let st = self.state.lock().unwrap();
            self.schedules = st.schedules.clone();
            self.bypass_schedule = st.bypass_schedule.clone();
        }

        let now_epoch = unsafe { esp_idf_svc::sys::time(std::ptr::null_mut()) } as i64;
        if now_epoch < 1_700_000_000 {
            return; // NTP not synced yet
        }

        // Get local time
        let mut tm: esp_idf_svc::sys::tm = unsafe { std::mem::zeroed() };
        let time_t = now_epoch as esp_idf_svc::sys::time_t;
        unsafe { esp_idf_svc::sys::localtime_r(&time_t, &mut tm) };

        let hour = tm.tm_hour as u8;
        let minute = tm.tm_min as u8;
        let current_minutes = hour as u16 * 60 + minute as u16;
        // tm_wday: 0=Sun, convert to bitmask: bit0=Mon..bit6=Sun
        let day_bit: u8 = match tm.tm_wday {
            0 => 1 << 6, // Sunday
            d => 1 << (d - 1),
        };

        // Ventilation schedule
        if !self.timed_off_active {
            let match_index = self.evaluate_ventilation(current_minutes, day_bit);

            if match_index != self.last_active_index {
                self.last_active_index = match_index;
            }

            let mut st = self.state.lock().unwrap();
            if !st.schedule_override {
                if match_index >= 0 {
                    st.schedule_active = true;
                    let level = self.schedules[match_index as usize].vent_level;
                    if level != st.requested_vent_level {
                        st.requested_vent_level = level;
                        info!("Scheduler: Level set to {}", level);
                    }
                } else if st.schedule_active {
                    st.schedule_active = false;
                    st.requested_vent_level = st.config.ventilation_level;
                    info!("Scheduler: No schedule active, using default");
                }
            }
        }

        // Bypass schedule
        if self.bypass_schedule.enabled {
            let day = tm.tm_mday as u8;
            let month = (tm.tm_mon + 1) as u8;
            let is_summer = is_date_in_range(
                day, month,
                self.bypass_schedule.start_day, self.bypass_schedule.start_month,
                self.bypass_schedule.end_day, self.bypass_schedule.end_month,
            );
            let mut st = self.state.lock().unwrap();
            if is_summer != st.requested_bypass_open {
                st.requested_bypass_open = is_summer;
                info!("Scheduler: Bypass {}", if is_summer { "open (summer)" } else { "closed (winter)" });
            }
        }
    }

    fn evaluate_ventilation(&self, current_minutes: u16, day_bit: u8) -> i32 {
        for (i, entry) in self.schedules.iter().enumerate() {
            if !entry.enabled || (entry.active_days & day_bit) == 0 {
                continue;
            }
            let start = entry.start_hour as u16 * 60 + entry.start_minute as u16;
            let end = entry.end_hour as u16 * 60 + entry.end_minute as u16;
            if is_time_in_range(current_minutes, start, end) {
                return i as i32;
            }
        }
        -1
    }

    pub fn activate_timed_off(&mut self, hours: u8) {
        if hours == 0 || hours > 99 { return; }
        let now_epoch = unsafe { esp_idf_svc::sys::time(std::ptr::null_mut()) } as i64;
        if now_epoch < 1_700_000_000 { return; }

        self.timed_off_active = true;
        self.timed_off_end_epoch = now_epoch + (hours as i64) * 3600;

        let mut st = self.state.lock().unwrap();
        st.requested_vent_level = 0; // Off
        st.schedule_override = true;
        st.timed_off_end_epoch = self.timed_off_end_epoch;
        st.persist_timed_off = true;
        info!("Scheduler: Timed off for {}h", hours);
    }

    pub fn cancel_timed_off(&mut self) {
        self.timed_off_active = false;
        self.timed_off_end_epoch = 0;
        let mut st = self.state.lock().unwrap();
        st.requested_vent_level = 2; // Normal
        st.schedule_override = false;
        st.timed_off_end_epoch = 0;
        st.persist_timed_off = true;
        info!("Scheduler: Timed off cancelled");
    }

    pub fn timed_off_remaining_minutes(&self) -> u32 {
        if !self.timed_off_active { return 0; }
        let now = unsafe { esp_idf_svc::sys::time(std::ptr::null_mut()) } as i64;
        let remaining = self.timed_off_end_epoch - now;
        if remaining > 0 { (remaining / 60) as u32 } else { 0 }
    }
}

fn is_time_in_range(current: u16, start: u16, end: u16) -> bool {
    if start <= end {
        current >= start && current < end
    } else {
        // Crosses midnight
        current >= start || current < end
    }
}

fn is_date_in_range(day: u8, month: u8, start_day: u8, start_month: u8, end_day: u8, end_month: u8) -> bool {
    let current = (month as u16) * 100 + day as u16;
    let start = (start_month as u16) * 100 + start_day as u16;
    let end = (end_month as u16) * 100 + end_day as u16;
    if start <= end {
        current >= start && current <= end
    } else {
        current >= start || current <= end
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_time_in_range_normal() {
        assert!(is_time_in_range(600, 480, 720)); // 10:00 in 8:00-12:00
        assert!(!is_time_in_range(300, 480, 720)); // 5:00 not in 8:00-12:00
    }

    #[test]
    fn test_time_in_range_midnight() {
        assert!(is_time_in_range(1400, 1320, 360)); // 23:20 in 22:00-06:00
        assert!(is_time_in_range(120, 1320, 360));   // 02:00 in 22:00-06:00
        assert!(!is_time_in_range(600, 1320, 360));  // 10:00 not in 22:00-06:00
    }

    #[test]
    fn test_date_in_range_normal() {
        assert!(is_date_in_range(15, 6, 15, 4, 30, 9)); // Jun 15 in Apr 15 - Sep 30
        assert!(!is_date_in_range(1, 2, 15, 4, 30, 9)); // Feb 1 not in Apr 15 - Sep 30
    }

    #[test]
    fn test_date_in_range_year_wrap() {
        assert!(is_date_in_range(15, 11, 1, 10, 31, 3)); // Nov 15 in Oct 1 - Mar 31
        assert!(is_date_in_range(15, 2, 1, 10, 31, 3));   // Feb 15 in Oct 1 - Mar 31
        assert!(!is_date_in_range(15, 6, 1, 10, 31, 3));  // Jun 15 not in Oct 1 - Mar 31
    }
}
